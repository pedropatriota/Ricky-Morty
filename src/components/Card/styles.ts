import styled from 'styled-components';

export const Container = styled.div`
	display: flex;
	flex-direction: column;
	width: 18rem;
`;

export const ImagContainer = styled.div<{ image: string }>`
	background-image: url(${props => props.image});
	background-repeat: no-repeat;
	background-size: cover;
	width: 100%;
	max-width: 18rem;
	height: 18rem;
	border-radius: 1rem 1rem 0 0;
`;

export const InfoContainer = styled.div`
	padding: 0.5rem;
	background-color: ${props => props.theme.colors.el};
	border: 1px solid ${props => props.theme.colors.el};
	border-radius: 0 0 1rem 1rem;

	h2 {
		margin-bottom: 1rem;
		color: ${props => props.theme.colors.text};
	}
	p {
		margin-bottom: 0.4rem;
		color: ${props => props.theme.colors.text};
	}
`;
