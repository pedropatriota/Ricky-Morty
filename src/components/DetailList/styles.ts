import styled from 'styled-components';

export const Container = styled.div`
	display: flex;
	gap: 0.75rem;
	flex-direction: column;
	margin-bottom: 0.75rem;
`;
