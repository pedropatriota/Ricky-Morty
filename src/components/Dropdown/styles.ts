import styled from 'styled-components';
import Select from 'react-select';

export const ReactSelect = styled(Select)`
  .react-select__control {
    width: min(90vw, 12.5rem);
    background-color: ${(props) => props.theme.colors.el};
    border-color: ${(props) => props.theme.colors.border};
    box-shadow: 0 0 0 transparent;
    height: 3.375rem;

    &:hover {
      border-color: ${(props) => props.theme.colors.border};
    }
  }

  .react-select__single-value {
    color: ${(props) => props.theme.colors.text};
  }

  .react-select__menu {
    width: 12.5rem;
    background-color: ${(props) => props.theme.colors.el};
    color: ${(props) => props.theme.colors.text};
  }

  .react-select__option,
  .react-select__option--is-focused 
  {
    background-color: ${(props) => props.theme.colors.el};
    color: ${(props) => props.theme.colors.text};
  }

  .react-select__dropdown-indicator,
  .react-select__indicator-separator {
    display: none;
  }
`;
