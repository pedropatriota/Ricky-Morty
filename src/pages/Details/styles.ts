import styled from 'styled-components';

export const Title = styled.h1`
	margin-bottom: '1.5rem';
`;

export const GoBackIcon = styled.img`
	width: 100%;
	max-width: 19rem;
	height: auto;
`;

export const InfoContainer = styled.div`
	display: flex;
	gap: 2rem;
	align-items: center;
	flex-wrap: wrap;
`;
