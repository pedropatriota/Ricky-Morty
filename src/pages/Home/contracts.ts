import { SingleValue } from 'react-select';

export type TSelected = SingleValue<{ label: string | null; value: string | null }>;

export interface ISelect {
	gender: TSelected;
	status: TSelected;
}

export type TOrder = 'Ascending' | 'Descending' | undefined;
