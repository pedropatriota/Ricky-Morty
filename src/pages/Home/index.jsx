/* eslint-disable react/jsx-no-comment-textnodes */
import useHome from './use-home';
import { Filter, List, Template, Order } from '../../components';

const Home = () => {
	const {
		allData,
		fetchNextPage,
		hasNextPage,
		isLoading,
		isError,
		error,
		select,
		inputValue,
		handleInputChange,
		handleFilterGender,
		handleFilterStatus,
		genderOptions,
		statusOptions,
		handleOrder,
		order
	} = useHome();

	if (isLoading) return <div>Loading...</div>;
	if (isError) return <div>{error?.toString()}</div>;

	return (
		<Template>
			<Order order={order} handleOrder={handleOrder} />
			<Filter
				handleChange={handleInputChange}
				handleFilter={[handleFilterGender, handleFilterStatus]}
				value={inputValue}
				filter={[select?.gender, select?.status]}
				options={[genderOptions, statusOptions]}
				label={['gender', 'status']}
			/>

			<List
				allData={allData}
				fetchNextPage={fetchNextPage}
				hasNextPage={hasNextPage}
			/>
		</Template>
	);
};

export default Home;
