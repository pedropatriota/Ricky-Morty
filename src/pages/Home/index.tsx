import useHome from './use-home';
import InfiniteScroll from 'react-infinite-scroller';
import { Character, Filter } from '../../components';

const Home = () => {
	const {
		allData,
		fetchNextPage,
		hasNextPage,
		isLoading,
		isError,
		error,
		select,
		inputValue,
		handleChange,
		handleSelect,
		options
	} = useHome();

	if (isLoading) return <div>Loading...</div>;
	if (isError) return <div>{(error as any)?.toString()}</div>;

	return (
		<>
			<Filter
				filter={select}
				handleChange={handleChange}
				handleFilter={handleSelect}
				value={inputValue}
				options={options}
			/>

			<InfiniteScroll loadMore={() => fetchNextPage()} hasMore={hasNextPage}>
				{allData.pages.map(({ results }) =>
					results.map(({ id, ...props }: any) => (
						<Character key={id} {...props} />
					))
				)}
			</InfiniteScroll>
		</>
	);
};

export default Home;
