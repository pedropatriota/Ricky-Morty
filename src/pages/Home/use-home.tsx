import { startTransition, useCallback, useMemo, useState, useContext } from 'react';
import { FilterContext } from '../../context/filter';
import { useLoadMoreAllData } from '../../service/hooks';
import type { TOrder, TSelected } from './contracts';

const useHome = () => {
	const RESOURCE = 'character';

	const {
		filters,
		setFilters,
		select,
		setSelect,
		inputValue,
		setInputValue,
		order,
		setOrder
	} = useContext(FilterContext);

	const { allData, fetchNextPage, hasNextPage, isLoading, isError, error } =
		useLoadMoreAllData({
			resource: RESOURCE,
			filters,
			order
		});

	const genderOptions = useMemo(
		() => [
			{ label: 'Female', value: 'Female' },
			{ label: 'Male', value: 'Male' },
			{ label: 'Unknown', value: 'unknown' }
		],
		[]
	);

	const statusOptions = useMemo(
		() => [
			{ label: 'Alive', value: 'Alive' },
			{ label: 'Dead', value: 'Dead' },
			{ label: 'Unknown', value: 'unknown' }
		],
		[]
	);

	const handleOrder = useCallback(() => {
		setOrder((prevOrder: TOrder): TOrder => {
			if (!prevOrder) return 'Ascending';
			else if (prevOrder === 'Ascending') return 'Descending';
			else return undefined;
		});
	}, [setOrder]);

	const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		event.preventDefault();
		setInputValue(event.target.value);
		startTransition(() => setFilters({ ...filters, name: event.target.value }));
	};

	const handleSelect = useCallback(
		(newValue: TSelected | any, propertyName: string) => {
			setSelect({ ...select, [propertyName]: newValue });
			startTransition(() =>
				setFilters({ ...filters, [propertyName]: newValue?.value as string })
			);
		},
		[filters, select, setFilters, setSelect]
	);

	return {
		allData,
		fetchNextPage,
		hasNextPage,
		isLoading,
		isError,
		resource: RESOURCE,
		error,
		genderOptions,
		statusOptions,
		inputValue,
		handleInputChange,
		handleFilterGender: (value: TSelected | any) => handleSelect(value, 'gender'),
		handleFilterStatus: (value: TSelected | any) => handleSelect(value, 'status'),
		handleSelect,
		select,
		handleOrder,
		order
	};
};

export default useHome;
