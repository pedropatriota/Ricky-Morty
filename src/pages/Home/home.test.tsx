import { renderWithProviders } from '../../utils';
import { cleanup, fireEvent, screen, waitFor } from '@testing-library/react';
import Home from '.';
import { light } from '../../styles/themes';

afterEach(() => cleanup);
describe('Home', () => {
	it('should fetch the API and render correctly', async () => {
		renderWithProviders(<Home />, light);

		const showNames = await screen.findAllByRole('heading');

		expect(showNames.length).toEqual(20);
	});

	it('should filter list when it types in the input', async () => {
		renderWithProviders(<Home />, light);

		const input = await screen.findByRole('textbox');

		fireEvent.change(input, { target: { value: 'Morty' } });

		const showNames = (await screen.findAllByRole('heading', {
			name: /Morty Smith|Alien Morty|Antenna Morty/i
		})) as HTMLElement[];

		expect(showNames).toHaveLength(3);
	});

	it('should filter on select a gender', async () => {
		renderWithProviders(<Home />, light);

		const selector = (await screen.findByLabelText(/gender/)) as Element;

		fireEvent.keyDown(selector, {
			key: 'ArrowDown',
			keyCode: 40,
			code: 40
		});

		const selectedItem = screen.getAllByText('Female');

		fireEvent.click(selectedItem[0]);

		const showNames = await screen.findAllByRole('heading', {
			name: /Summer Smith|Beth Smith|Abadango Cluster Princess|Annie/i
		});

		expect(showNames).toHaveLength(4);
	});

	it('should filter on select a status', async () => {
		renderWithProviders(<Home />, light);

		const selector = (await screen.findByLabelText(/status/)) as Element;

		fireEvent.keyDown(selector, {
			key: 'ArrowDown',
			keyCode: 40,
			code: 40
		});

		const selectedItem = screen.getAllByText('Dead');

		fireEvent.click(selectedItem[0]);

		const showNames = await screen.findAllByRole('heading', {
			name: /Adjudicator Rick|Agency Director|Alan Rails|Albert Einstein/i
		});

		expect(showNames).toHaveLength(4);
	});

	it('should order the list', async () => {
		renderWithProviders(<Home />, light);

		expect(await screen.findByText(/add order/i)).toBeTruthy();

		const button = await screen.findByRole('button', {
			name: /add order/i
		});

		fireEvent.click(button);

		expect(await screen.findByText(/ascending/i)).toBeTruthy();

		const showNames = await screen.findAllByRole('heading', {
			name: /Abadango Cluster Princess|Abradolf Lincler|Adjudicator Rick|Agency Director/i
		});

		expect(showNames).toHaveLength(4);

		fireEvent.click(button);

		expect(await screen.findByText(/descending/i)).toBeTruthy();

		const showNamesDescending = await screen.findAllByRole('heading', {
			name: /Summer Smith|Rick Sanchez|Morty Smith|Jerry Smith/i
		});

		expect(showNamesDescending).toHaveLength(4);
	});
});
