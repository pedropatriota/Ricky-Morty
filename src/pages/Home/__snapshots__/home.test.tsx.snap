// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should match the snapshot 1`] = `
<div>
  <div
    class="sc-jlZhew cYsoQy"
  >
    <style
      data-testid="loading-spin-style"
    >
      
                @keyframes spin {
                    0% {
                        transform: rotate(0deg);
                    }
                    100% {
                        transform: rotate(720deg);
                    }
                }
              
    </style>
    <div
      data-testid="loading-spin"
      style="display: inline-block; border-style: solid; border-radius: 50%; animation-iteration-count: infinite; animation-name: spin; height: 60px; width: 60px; border-width: 6px; animation-timing-function: ease-in-out; animation-direction: normal; animation-duration: 2s; border-color: #cccccc; border-left-color: #1ecd96; border-top-color: #1ecd96;"
    />
  </div>
</div>
`;
